client/hwlib/src/collectors/cpuid.rs

```diff
@@ -0,0 +1,128 @@
+/* Copyright 2024 Canonical Ltd.
+ * All rights reserved.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+use anyhow::Result;
+use itertools::Itertools;
+use smbioslib;
+use std::fmt::{self, Display};
+
+#[derive(Clone, Debug)]
+pub struct CpuId(String);
+
+impl CpuId {
+    pub fn new(proc_info: &smbioslib::SMBiosProcessorInformation) -> Result<Self> {
+        let cpu_id = proc_info
+            .processor_id()
+            .map(|id| format!("{:x?}", id))
+            .ok_or_else(|| anyhow::anyhow!("Processor ID not found"))?;
+        Ok(CpuId(cpu_id))
+    }
+
+    pub fn codename(&self) -> Option<String> {
+        if self.0.is_empty() {
+            return None;
+        }
+
+        // Collect the first three whitespace-separated parts into a tuple
+        let cpu_id_parts = self
+            .0
+            .split_whitespace()
+            .take(3)
+            .collect_tuple::<(_, _, _)>();
+
+        // CPU ID bytes are represented as strings
+        if let Some((byte1, byte2, byte3)) = cpu_id_parts {
+            // Reverse the order of the bytes and create the hex string
+            let cpu_id_hex = format!("0x{}{}{}", byte3, byte2, byte1);
+
+            // Translate the CPU ID into a human-friendly name
+            Some(
+                self.to_human_friendly(&cpu_id_hex)
+                    .unwrap_or_default()?
+                    .to_string(),
+            )
+        } else {
+            None
+        }
+    }
+
+    /// Implement the same logic as used in C3 to get CPU codename
+    /// https://github.com/canonical/hexr/blob/0d6726f00f9fa77efdae201188ad10a8bbbfb2be/apps/hardware/parsers/cpuid.py#L29
+    fn to_human_friendly(&self, cpu_id_hex: &str) -> Result<Option<&'static str>> {
```

Comment:
I think if we use `u64` here, we'll have to deal with reversing it. Technically strings like `0x806e9` represent hex numbers, and that's why we ignore case, so it would be more natural to represent them as `u64`. However, I'm concerned that we'll have to deal with multiple conversions between `str` and `u64`, so I guess it's easier to use strings here

Reply:
I've also updated it to return `Option` in 522a5bb

Reply:
That said, if this is something `smbios` is already doing, consistency might be okay

Reply:
I'll update it and add the proper tests to make sure that we don't mess up the logic

Reply:
Updated in 37526e6